File: tests/transpile/probes/_scripts/_reshape.py
"""Shared in-container reshape helper for probe scripts.

The host harness sends each (param, data) value as a flat
row-major Python list under `points.json`. The model the probe
loads, however, typically declares the same name with a multi-
dimensional declared shape (e.g. Stan `array[16] int obs` or PyMC
`obs.shape == (16, 4)`). Each probe script imports
:func:`reshape_point` and rebuilds the multi-dimensional native
container the underlying runtime expects.

The reshape table arrives in `/io/shapes.json`:

    {"obs": [16], "rule_weights": [16], "logits": [3, 4]}

The dtype table (optional) arrives in `/io/dtypes.json`:

    {"obs": "int", "logits": "float"}

A name absent from either table falls through to its raw
list value. The float / int cast applies even for scalars so
backends that distinguish `int` and `real` (Stan, JAGS) get the
right type.

The export channel rides alongside. `/io/export_names.json` holds
the QVR program's return-variable names, in declaration order:

    ["phi"]

A probe that finds the file must read the exported value out of the
emitted program's own return surface (the model function's `return`,
a Stan generated quantity, a Gen `assess` return value) and report
one entry per name per point under the result's `exports` key. The
file's absence means the caller did not ask for the export channel,
and the probe reports only log-densities.
"""
from __future__ import annotations

import json
import pathlib
import re
from collections.abc import Sequence
from typing import TYPE_CHECKING, cast

# The probe payload is strictly numeric: scalars (Python int or
# float) and arbitrarily nested lists of the same. Aliases live
# under `TYPE_CHECKING` so the helper imports on every in-container
# Python (3.9 in JAGS / BUGS, 3.11 in Edward2 / Node, 3.12 in the
# rest); `from __future__ import annotations` defers annotation
# evaluation so the recursive `NestedNumber` reference resolves
# lazily for type checkers and never executes at runtime. `Any`
# and `object` are intentionally avoided.
if TYPE_CHECKING:
    from typing import Protocol

    Number = int | float
    NestedNumber = Number | list["NestedNumber"]
    PointSection = dict[str, NestedNumber]
    Point = dict[str, PointSection]

    class ArrayLike(Protocol):
        """Any runtime array exposing numpy's `tolist`.

        Every probe runtime that carries an exported value back to
        the harness (numpy, jax, torch, tensorflow) satisfies this,
        so the conversion needs neither a runtime import of those
        libraries nor a bare `object` annotation.
        """

        def tolist(self) -> NestedNumber:
            ...

    ExportValue = NestedNumber | ArrayLike | tuple["ExportValue", ...]


def load_tables(io: pathlib.Path) -> tuple[
    dict[str, list[int]], dict[str, str],
]:
    """Read `/io/shapes.json` and `/io/dtypes.json`. Each absent
    file resolves to an empty dict so the helper is a no-op in
    legacy probe runs that don't ship the tables."""
    shapes_path = io / "shapes.json"
    dtypes_path = io / "dtypes.json"
    shapes = (
        json.loads(shapes_path.read_text())
        if shapes_path.exists() else {}
    )
    dtypes = (
        json.loads(dtypes_path.read_text())
        if dtypes_path.exists() else {}
    )
    return shapes, dtypes


def _flat_to_nested(
    flat: Sequence[Number], shape: list[int],
) -> NestedNumber:
    """Reshape a flat row-major list into a nested list with the
    given shape. Returns the scalar element when `shape == []`."""
    if not shape:
        if len(flat) != 1:
            raise ValueError(
                f"scalar shape but len(flat)={len(flat)}"
            )
        return flat[0]
    expected = 1
    for d in shape:
        expected *= d
    if len(flat) != expected:
        raise ValueError(
            f"flat length {len(flat)} does not match shape "
            f"{shape} (expected {expected})"
        )
    if len(shape) == 1:
        return list(flat)
    stride = expected // shape[0]
    return [
        _flat_to_nested(flat[i * stride:(i + 1) * stride], shape[1:])
        for i in range(shape[0])
    ]


def reshape_value(
    name: str,
    value: NestedNumber,
    shapes: dict[str, list[int]],
    dtypes: dict[str, str],
) -> NestedNumber:
    """Reshape one (name, value) pair using the loaded tables.

    - When `name` has a shape entry, treat `value` as flat row-major
      and rebuild a nested list of the declared shape (scalar shape
      `[]` collapses a length-1 list to its element).
    - When `name` has a dtype entry, recursively cast every leaf to
      `int` or `float`.
    - When neither table mentions `name`, return the value unchanged
      (preserves the existing 7-fixture probe contract).
    """
    if name in shapes:
        if isinstance(value, (int, float)):
            value = _flat_to_nested([value], shapes[name])
        elif isinstance(value, list):
            value = _flat_to_nested(
                cast("Sequence[Number]", value), shapes[name],
            )
    if name in dtypes:
        value = _cast_leaves(value, dtypes[name])
    return value


def _cast_leaves(value: NestedNumber, dtype: str) -> NestedNumber:
    """Recursively cast every leaf of a nested list to the target
    dtype (`"int"` -> Python int, anything else -> float)."""
    if isinstance(value, list):
        return [_cast_leaves(v, dtype) for v in value]
    if dtype == "int":
        return int(value)
    return float(value)


def reshape_point(
    point: Point,
    shapes: dict[str, list[int]],
    dtypes: dict[str, str],
) -> Point:
    """Reshape every entry under `params` / `data` in one point."""
    out: Point = {}
    for section in ("params", "data"):
        out[section] = {
            name: reshape_value(name, value, shapes, dtypes)
            for name, value in point.get(section, {}).items()
        }
    return out


def index_input_names(
    source: str, dtypes: dict[str, str],
) -> set[str]:
    """Names the emitted source uses as array subscripts.

    The gallery datasets index every plate with a 0-based integer
    covariate (``out_idx``, ``cat_idx``, ``word_idx``): the QVR
    program and the row-major `Point` payload both count from 0. A
    backend whose native array indexing counts from 1 (Stan, JAGS,
    BUGS, and the Julia targets Turing / Gen) subscripts the gathered
    parameter with that covariate directly, so the host must lift
    every such covariate to 1-based before handing it to the
    container.

    An index input is any ``int``-dtyped name the source subscripts,
    i.e. one that appears immediately after a ``[``. Integer values
    that are never subscripts (count observations such as ``tally`` /
    ``obs``, Bernoulli 0/1 responses) are left untouched: they are
    outcomes, not offsets.
    """
    names: set[str] = set()
    for name, dtype in dtypes.items():
        if dtype != "int":
            continue
        if re.search(r"\[\s*" + re.escape(name) + r"(?![0-9A-Za-z_])", source):
            names.add(name)
    return names


def _offset_leaves(value: NestedNumber, offset: int) -> NestedNumber:
    """Recursively add ``offset`` to every leaf of a nested list."""
    if isinstance(value, list):
        return [_offset_leaves(v, offset) for v in value]
    return value + offset


def shift_index_inputs(
    point: Point, names: set[str], offset: int = 1,
) -> Point:
    """Return ``point`` with every ``names`` entry's leaves shifted.

    Applied by the 1-based backends after :func:`reshape_point` so a
    0-based covariate becomes a valid 1-based subscript. Names outside
    ``names`` pass through unchanged, so count observations and
    response values keep their raw magnitude."""
    out: Point = {}
    for section in ("params", "data"):
        out[section] = {
            name: (
                _offset_leaves(value, offset) if name in names else value
            )
            for name, value in point.get(section, {}).items()
        }
    return out


def load_export_names(io: pathlib.Path) -> list[str]:
    """Read `/io/export_names.json`.

    Returns the QVR program's return-variable names in declaration
    order, or an empty list when the caller did not ship the file.
    An empty list means "do not report the export channel"; it never
    means "the program exports nothing", because a program with no
    return clause is never scheduled through this channel in the
    first place.
    """
    path = io / "export_names.json"
    if not path.exists():
        return []
    return [str(name) for name in json.loads(path.read_text())]


def as_nested(value: ExportValue) -> NestedNumber:
    """Convert one runtime array / scalar into JSON-ready numbers.

    Every probe runtime that can carry an exported value back to the
    harness (numpy, jax, torch, tensorflow) exposes numpy's `tolist`,
    so the conversion dispatches on that method rather than on the
    concrete array class. A Python tuple (the shape a multi-name
    return takes in every Python target) recurses elementwise; a bare
    scalar widens to `float`.

    A boolean leaf stays an `int`: a Bernoulli-supported export is
    integer-valued, and JSON's `true` would compare unequal against
    the reference's `1.0` under a numeric tolerance.

    Three attribute probes cover every runtime the images carry, in
    order of directness: `tolist` (numpy, jax, torch), `numpy` (a
    TensorFlow eager tensor, which has no `tolist`), and `value`
    (an Edward2 `RandomVariable`, which wraps the tensor its export
    denotes).
    """
    to_list = getattr(value, "tolist", None)
    if to_list is not None:
        return to_list()
    to_numpy = getattr(value, "numpy", None)
    if to_numpy is not None:
        return as_nested(to_numpy())
    wrapped = getattr(value, "value", None)
    if wrapped is not None:
        return as_nested(wrapped)
    if isinstance(value, tuple):
        return [as_nested(item) for item in value]
    if isinstance(value, list):
        return [as_nested(item) for item in value]
    if isinstance(value, bool):
        return int(value)
    if isinstance(value, int):
        return value
    return float(value)


def export_payload(
    names: list[str], returned: ExportValue | None,
) -> list[NestedNumber]:
    """Split one model return value into one entry per export name.

    A program returning a single name hands back a bare value; a
    program returning several hands back a tuple, in the order the
    `return` clause declares. The arity is checked rather than
    assumed: a return whose arity disagrees with the requested names
    is a renderer defect (a dropped or duplicated export), and it has
    to surface as a probe failure rather than as a silently truncated
    comparison.
    """
    if not names:
        raise ValueError(
            "export_payload called with no export names; the caller "
            "did not ship /io/export_names.json and the probe must "
            "not report an export channel."
        )
    if returned is None:
        raise ValueError(
            f"the emitted model returns nothing where the QVR program "
            f"exports {names}. A transpilation that drops the return "
            f"clause emits a program denoting the right joint and the "
            f"wrong kernel."
        )
    if len(names) == 1:
        return [as_nested(returned)]
    if not isinstance(returned, tuple):
        raise ValueError(
            f"the emitted model returns a single value where the QVR "
            f"program exports {len(names)} ({names}). The renderer "
            f"dropped part of the program's return clause."
        )
    if len(returned) != len(names):
        raise ValueError(
            f"the emitted model returns {len(returned)} value(s) "
            f"where the QVR program exports {len(names)} ({names})."
        )
    return [as_nested(item) for item in returned]


__all__ = [
    "as_nested",
    "export_payload",
    "index_input_names",
    "load_export_names",
    "load_tables",
    "reshape_point",
    "reshape_value",
    "shift_index_inputs",
]

File: tests/transpile/probes/_scripts/_reshape.jl
# Shared in-container reshape helper for Julia probes.
#
# Loads `/io/shapes.json` and `/io/dtypes.json` (each absent file
# resolves to an empty dict so legacy probe runs are a no-op),
# rebuilds nested arrays from row-major flat lists, and casts
# leaves to Int / Float64 per dtype.
#
# `reshape_value` is the per-name entry point; `reshape_point`
# takes a JSON3.Object with `params` / `data` sections and returns
# a `Dict{Symbol,Any}` ready for the runtime-specific coercion in
# the calling probe (turing.jl, gen.jl).

using JSON3

function _load_json_dict(path::AbstractString)
    if !isfile(path)
        return Dict{String,Any}()
    end
    raw = JSON3.read(read(path, String))
    # JSON3.Object iterates as (Symbol, value); rebuild with
    # explicitly stringified keys so the downstream
    # `Dict{String,Vector{Int}}` / `Dict{String,String}` typed
    # conversions accept the entries.
    return Dict{String,Any}(String(k) => v for (k, v) in pairs(raw))
end

function load_tables(io::AbstractString)
    shapes_raw = _load_json_dict(joinpath(io, "shapes.json"))
    dtypes_raw = _load_json_dict(joinpath(io, "dtypes.json"))
    shapes = Dict{String,Vector{Int}}(
        k => [Int(x) for x in v] for (k, v) in shapes_raw
    )
    dtypes = Dict{String,String}(k => String(v) for (k, v) in dtypes_raw)
    return shapes, dtypes
end

function _flat_to_nested(flat::AbstractVector, shape::Vector{Int})
    if isempty(shape)
        length(flat) == 1 || error(
            "scalar shape but length(flat)=$(length(flat))"
        )
        return flat[1]
    end
    expected = prod(shape)
    length(flat) == expected || error(
        "flat length $(length(flat)) does not match shape $shape "
        * "(expected $expected)"
    )
    if length(shape) == 1
        return collect(flat)
    end
    stride_ = div(expected, shape[1])
    return [
        _flat_to_nested(flat[(i - 1) * stride_ + 1:i * stride_], shape[2:end])
        for i in 1:shape[1]
    ]
end

function _cast_leaves(value, dtype::String)
    if value isa AbstractArray
        return [_cast_leaves(v, dtype) for v in value]
    end
    return dtype == "int" ? Int(value) : Float64(value)
end

function reshape_value(
    name::String, value, shapes::Dict, dtypes::Dict,
)
    out = value
    if haskey(shapes, name)
        if out isa Real
            out = _flat_to_nested([out], shapes[name])
        elseif out isa AbstractArray
            out = _flat_to_nested(collect(out), shapes[name])
        elseif out isa JSON3.Array
            out = _flat_to_nested([x for x in out], shapes[name])
        end
    end
    if haskey(dtypes, name)
        out = _cast_leaves(out, dtypes[name])
    end
    return out
end

function reshape_point(pt, shapes::Dict, dtypes::Dict)
    out = Dict{String,Dict{Symbol,Any}}()
    for section in ("params", "data")
        sec = haskey(pt, section) ? pt[section] : Dict{Symbol,Any}()
        out[section] = Dict{Symbol,Any}(
            Symbol(k) => reshape_value(String(k), v, shapes, dtypes)
            for (k, v) in sec
        )
    end
    return out
end

# `reshape_value` returns a nested `Vector` of `Vector`s for a rank >= 2
# name, which is the shape a JSON payload naturally rebuilds into but
# not one Distributions.jl or Gen accepts. `native_array` projects that
# nesting onto a dense `Array{T,N}` with the same axis order. Julia is
# column-major and the wire payload is row-major, so the leaves reshape
# against the reversed dimension list and then permute back.
function nested_shape(value)
    dims = Int[]
    cur = value
    while cur isa AbstractArray
        push!(dims, length(cur))
        isempty(cur) && break
        cur = first(cur)
    end
    return dims
end

function _collect_leaves!(out, value)
    if value isa AbstractArray
        for x in value
            _collect_leaves!(out, x)
        end
    else
        push!(out, value)
    end
    return out
end

function nested_leaves(value)
    return _collect_leaves!(Any[], value)
end

function native_array(value)
    if !(value isa AbstractArray)
        if value isa Integer
            return Int(value)
        elseif value isa Real
            return Float64(value)
        end
        return value
    end
    if ndims(value) > 1
        # Already a dense multi-dimensional array; only the element
        # type needs projecting.
        elt = all(x -> x isa Integer, value) ? Int : Float64
        return Array{elt}(value)
    end
    dims = nested_shape(value)
    leaves = nested_leaves(value)
    elt = all(x -> x isa Integer, leaves) ? Int : Float64
    typed = elt[elt(x) for x in leaves]
    length(dims) <= 1 && return typed
    return permutedims(
        reshape(typed, reverse(dims)...), collect(length(dims):-1:1),
    )
end

# The Julia targets (Turing, Gen) index arrays from 1, but the gallery
# covariates count from 0. `index_input_names` finds every int-dtyped
# name the source subscripts (`[name`); `shift_index_inputs` lifts
# those entries to 1-based after reshape. Count observations and
# response values are never subscripts, so they pass through untouched.
function index_input_names(source::AbstractString, dtypes::Dict)
    names = Set{String}()
    for (name, dt) in dtypes
        if dt == "int" && occursin(
            Regex("\\[\\s*" * name * "(?![0-9A-Za-z_])"), source,
        )
            push!(names, name)
        end
    end
    return names
end

function _offset_leaves(value, offset::Int)
    if value isa AbstractArray
        return [_offset_leaves(v, offset) for v in value]
    end
    return value + offset
end

function shift_index_inputs(point, names, offset::Int = 1)
    out = Dict{String,Dict{Symbol,Any}}()
    for section in ("params", "data")
        sec = point[section]
        out[section] = Dict{Symbol,Any}(
            k => (String(k) in names ? _offset_leaves(v, offset) : v)
            for (k, v) in sec
        )
    end
    return out
end

# ---------------------------------------------------------------------
# The export channel.
#
# `/io/export_names.json` holds the QVR program's return-variable
# names, in declaration order. A probe that finds the file reads the
# exported value out of the emitted program's own return surface (the
# `@model` return under `condition`, the second element of
# `Gen.assess`) and reports one entry per name per point.
# ---------------------------------------------------------------------

function load_export_names(io::AbstractString)
    path = joinpath(io, "export_names.json")
    isfile(path) || return String[]
    return [String(x) for x in JSON3.read(read(path, String))]
end

# Nest a returned value row-major so JSON3 emits the same layout the
# host reference produced. Writing a Julia `Matrix` directly would
# serialise its column-major storage as one flat array, which compares
# element-for-element against a transposed reference.
function export_nested(v)
    if v isa AbstractArray
        if ndims(v) == 0
            return export_nested(v[])
        elseif ndims(v) == 1
            return [export_nested(x) for x in v]
        end
        rest = ntuple(_ -> Colon(), ndims(v) - 1)
        return [export_nested(v[i, rest...]) for i in axes(v, 1)]
    elseif v isa Tuple
        return [export_nested(x) for x in v]
    elseif v isa Bool
        return Int(v)
    elseif v isa Integer
        return Int(v)
    end
    return Float64(v)
end

function export_payload(names::Vector{String}, returned)
    isempty(names) && error(
        "export_payload called with no export names; the caller did " *
        "not ship /io/export_names.json and the probe must not " *
        "report an export channel."
    )
    returned === nothing && error(
        "the emitted model returns nothing where the QVR program " *
        "exports $(names). A transpilation that drops the return " *
        "clause emits a program denoting the right joint and the " *
        "wrong kernel."
    )
    if length(names) == 1
        return [export_nested(returned)]
    end
    returned isa Tuple || error(
        "the emitted model returns a single value where the QVR " *
        "program exports $(length(names)) ($(names)). The renderer " *
        "dropped part of the program's return clause."
    )
    length(returned) == length(names) || error(
        "the emitted model returns $(length(returned)) value(s) " *
        "where the QVR program exports $(length(names)) ($(names))."
    )
    return [export_nested(x) for x in returned]
end
